lib/puppet/property.rb

```diff
@@ -346,6 +362,21 @@ def insync?(is)
     end
   end
 
+  # This method tests if two values are insync? outside of the properties current
+  # should value. This works around the requirement for corrective_change analysis
+  # that requires two older values to be compared with the properties potentially
+  # custom insync? code.
+  #
+  # @param [Object] orig_should the original should value
+  # @param [Object] orig_is the original is value
+  # @return [Boolean] whether or not the values are in sync or not
+  # @api private
+  def insync_values?(orig_should, orig_is)
+    test_property = self.dup
```

Comment:
I would be happy for a dpittman essay, so it's clear to anyone that sees this that we don't want this behavior to replicate.